### PR TITLE
Add handling of negative counter

### LINKED ISSUE #583: Unrecognized GCOV output
**Describe the bug**
We have been experiencing parsing errors in our CI pipeline that utilizes gcovr for python for the past day.  This had not been happening for at least a few months prior since we incorporated gcovr.

**To Reproduce**
https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true
```
(WARNING) Unrecognized GCOV output for /home/runner/work/zlib-ng/zlib-ng/deflate_medium.c
	  branch  3 taken -7476[72](https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true#step:8:72)
	This is indicative of a gcov output parse error.
	Please report this to the gcovr developers
	at <https://github.com/gcovr/gcovr/issues>.
(WARNING) Exception during parsing:
	UnknownLineType: branch  3 taken -[74](https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true#step:8:74)[76](https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true#step:8:76)72
(ERROR) Exiting because of parse errors.
	You can run gcovr with --gcov-ignore-parse-errors
	to continue anyway.
...
(DEBUG)   Filtering coverage data for file /usr/include/x86_64-linux-gnu/bits/string_fortified.h
Traceback (most recent call last):
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line 77, in worker
    work(*args, **kwargs)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 317, in process_datafile
    done = run_gcov_and_process_files(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 490, in run_gcov_and_process_files
    process_gcov_data(fname, covdata, abs_filename, options)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line [145](https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true#step:8:145), in process_gcov_data
    coverage = parse_coverage(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 370, in parse_coverage
    _report_lines_with_errors(lines_with_errors, context)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 509, in _report_lines_with_errors
    raise errors[0]  # guaranteed to have at least one exception
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 322, in parse_coverage
    tokenized_lines.append((_parse_line(raw_line), raw_line))
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 777, in _parse_line
    raise UnknownLineType(line)
gcovr.gcov_parser.UnknownLineType: branch  3 taken -747672
Traceback (most recent call last):
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line 77, in worker
    work(*args, **kwargs)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 317, in process_datafile
    done = run_gcov_and_process_files(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 490, in run_gcov_and_process_files
    process_gcov_data(fname, covdata, abs_filename, options)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 145, in process_gcov_data
    coverage = parse_coverage(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 370, in parse_coverage
    _report_lines_with_errors(lines_with_errors, context)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 509, in _report_lines_with_errors
    raise errors[0]  # guaranteed to have at least one exception
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 322, in parse_coverage
    tokenized_lines.append((_parse_line(raw_line), raw_line))
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 777, in _parse_line
    raise UnknownLineType(line)
gcovr.gcov_parser.UnknownLineType: branch  3 taken -747672
(ERROR) Uncaught EXCEPTION
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/__main__.py", line 521, in <module>
    main()
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/__main__.py", line 308, in main
    collect_coverage_from_gcov(covdata, options)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/__main__.py", line 367, in collect_coverage_from_gcov
    contexts = pool.wait()
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line 170, in __exit__
    self.wait()
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line [161](https://github.com/zlib-ng/zlib-ng/runs/5710429496?check_suite_focus=true#step:8:161), in wait
    raise self.exceptions[0][1]
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/__main__.py", line 367, in collect_coverage_from_gcov
    contexts = pool.wait()
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line 161, in wait
    raise self.exceptions[0][1]
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/workers.py", line 77, in worker
    work(*args, **kwargs)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 317, in process_datafile
    done = run_gcov_and_process_files(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 490, in run_gcov_and_process_files
    process_gcov_data(fname, covdata, abs_filename, options)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov.py", line 145, in process_gcov_data
    coverage = parse_coverage(
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 370, in parse_coverage
    _report_lines_with_errors(lines_with_errors, context)
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 509, in _report_lines_with_errors
    raise errors[0]  # guaranteed to have at least one exception
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 322, in parse_coverage
    tokenized_lines.append((_parse_line(raw_line), raw_line))
  File "/home/runner/.local/lib/python3.8/site-packages/gcovr/gcov_parser.py", line 777, in _parse_line
    raise UnknownLineType(line)
gcovr.gcov_parser.UnknownLineType: branch  3 taken -747672
```

**Expected behavior**
No parser errors.

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - GCC version 9.3.0
 - GCOVR version 5.1

### PR DESCRIPTION
**Pull request by Spacetown**
Add option `--gcov-ignore-negative-counters` to ignore negative gcov counter values.
See https://github.com/gcovr/gcovr/issues/583#issuecomment-1340762818 for details.

Closes #583

### CODE REVIEW

**Comment by latk** [comment]
I'm simply not sure what the appropriate behaviour is here. Could you explain on a design level what the intended behaviour is, and why?

* what is the intended default behaviour?
* what is the intended behaviour if this flag is explicitly enabled/disabled?
* how does this interact with --gcov-ignore-parse-errors?
* what levels of control over the errors/warnings/messages is given to users?
  * can users get an error/warning for each occurrence of a problem?
  * can users get summarized errors/warnings so that they don't get flooded with messages?
  * can users completely silence any mention of this problem?

From what I understand:

* default behaviour: exception is raised to alert user to problem, this exception is subject to ignore-parse-error handling
* behaviour if flag is enabled: warning is logged, this warning does not interact with ignore-parse-error
* user will always get one message per occurrence of the error, no way to summarize or completely silence it

By summarization, I mean output such as the following that reports a warning once, and then reports how many instances of the warning type there were (without printing all of them):

```none
WARN foo.cpp:42: ignoring corrupted branch coverage data on gcov line "branch 4 taken -12345"
INFO see https://github.com/gcovr/gcovr/issue/... for details
WARN 500 duplicate warnings skipped
```

I am not saying that we should implement such summarization (it does involve tricky data flows), but want to make sure that “always show all warnings/errors” is the intended behaviour in all modes.

File: gcovr/tests/test_gcov_parser.py [hunk 1]
```diff
@@ -558,6 +564,68 @@ def test_branch_exclusion(flags):
...
+        parse_coverage(
+            source.splitlines(),
+            filename="example.cpp",
+            ignore_parse_errors=False,
+            ignore_negative_counters=False,
+        )
+    except Exception as e:
+        assert isinstance(e, NegativeCounter)
+    else:
+        assert False, "Here we should never come"
```
Comment (latk) [inline gcovr/tests/test_gcov_parser.py:593]: Could be simplified using `pytest.raises`:

```suggestion
    with pytest.raises(NegativeCounter):
        parse_coverage(
            source.splitlines(),
            filename="example.cpp",
            ignore_parse_errors=False,
            ignore_negative_counters=False,
        )
```

See: https://docs.pytest.org/en/7.2.x/how-to/assert.html#assertions-about-expected-exceptions

**Comment by Spacetown** [comment]
The intended default behavior for me is to reject the data and abort.

You're right that the user can get lots of warnings and a summarize would be helpful. Counting the issues is not so simple, since we need a parameter (summarize for the file) or a global value (summarize all occurrences).
The place where the error is logged is only aware of the parsed items and not the original line.
If the option is a enumeration we can use abort, log-all and log-first. In the last case we can add a message "... and possible more" instead of counting the errors. I think the user should be aware that there is a problem but if it occur 500 times doesn't give much more information.

Seeing your log example I think it would be better to check the value when splitting the line.

Another option would be to reuse the old flag. If we add an optional argument which defaults to `all` we can define which parser errors shall be ignored, e.g. `negative_counter[log_once]`.
This configuration can be described in deep in the documentation.

**Comment by Spacetown** [comment]
I've merged the option and we now only have the `--gcov-ignore-parse-errors` which got an optional argument. The option can be given several times and if no value is specified the value `all` is used.

The summary on file level is added but not on project level. I think it's important to warn for each file with unrecognized data and not only once per project.

File: doc/source/guide/gcov_parser.rst [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+Negative hit counts
+___________________
+
+A bug in gcov can produce negative hit values (see `gcov comment`) which are not accepted by default.
+This behavior can be changed by using the value ``--gcov-ignore-parse-errors=negative_hits.warn`` or
+``--gcov-ignore-parse-errors=negative_hits.warn_once_per_file``. The first form warns on every line
+with a negative value the second one only once per processed file and adds a summary with the overall
+issues in the file.
+
+.. _gcov comment https://github.com/gcovr/gcovr/issues/583#issuecomment-1340762818:
```
Comment (latk) [inline doc/source/guide/gcov_parser.rst:38]: The link syntax here is off. I think it should be:

```suggestion
A bug in gcov can produce negative hit values (see `gcov comment`_) which are not accepted by default.
This behavior can be changed by using the value ``--gcov-ignore-parse-errors=negative_hits.warn`` or
``--gcov-ignore-parse-errors=negative_hits.warn_once_per_file``. The first form warns on every line
with a negative value the second one only once per processed file and adds a summary with the overall
issues in the file.

.. _gcov comment: https://github.com/gcovr/gcovr/issues/583#issuecomment-1340762818
```

**Pull request merged**